src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,194 @@
+        if (isEmpty() || index == size) {
+            add(value);
+            return;
+        }
+
+        Node<T> currentNode = findNode(index);
+        Node<T> newNode = new Node<>(currentNode.prev, value, currentNode);
+
+        if (currentNode.equals(head)) {
+            head = newNode;
+        } else {
+            currentNode.prev.next = newNode;
+        }
+
+        currentNode.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element: list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = findNode(index);
+        return node.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = findNode(index);
+        T oldValue = currentNode.value;
+        currentNode.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> currentNode = findNode(index);
+        unlink(currentNode);
+        return currentNode.value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        Node<T> node = findNode(object);
+        if (node == null) {
+            return false;
+        }
+        unlink(node);
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndexForAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Can't add an element. Invalid index");
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Can't get an element. Invalid index");
+        }
+    }
+
+    private Node<T> findNode(int index) {
+        Node<T> currentNode;
+        if (index == 0 || size % index >= 2) {
+            currentNode = head;
+            for (int i = 0; i < index; i++) {
+                currentNode = currentNode.next;
+            }
+        } else {
+            currentNode = tail;
+            for (int i = 1; i < size - index; i++) {
+                currentNode = currentNode.prev;
+            }
+        }
+
+        return currentNode;
+    }
+
+    private Node<T> findNode(T object) {
+        Node<T> currentNode = head;
+        for (int i = 0; i < size; i++) {
+            if (currentNode.value == null && object == currentNode.value) {
+                return currentNode;
+            }
+            if (currentNode.value != null && currentNode.value.equals(object)) {
+                return currentNode;
+            }
+            currentNode = currentNode.next;
+        }
```

Comment:
If you use FOR - use it this way and think about one IF condition
```suggestion
        for (Node<T> node = head; node != null; node = node.next) {
            if (...) {
                return node ;
            }
        }
```